### PR TITLE
Enable to download and run snapshot version from dynamic

### PR DESCRIPTION
**Pull request by tanishiking**
Close https://github.com/scalameta/scalafmt/issues/1655

### Problem
As described in the issue, before this PR, scalafmt-dynamic (and so scalafmt-cli, sbt-scalafmt, every module that depends on scalafmt-dynamic) couldn't resolve the SNAPSHOT version of scalafmt-core.

- Run `sbt publishLocal` and publish the snapshot version (e.g. `2.2.2-SNAPSHOT`)
- Set `version` field in `.scalafmt.conf` to `2.2.2-SNAPSHOT`
- Try to run scalafmt using CLI / sbt-scalafmt

got

```
org.scalafmt.cli.FailedToFormat: /Users/tanishiking/dev/src/github.com/tanishiking/scalafmt/scalafmt-tests/src/test/scala/org/scalafmt/util/Tabulator.scala
Caused by: org.scalafmt.dynamic.exceptions.VersionMismatch: Scalafmt version mismatch. Version in .scalafmt.conf is '2.2.2-SNAPSHOT' and running version is '2.2.2'.
        at org.scalafmt.dynamic.ScalafmtReflect.checkVersionMismatch(ScalafmtReflect.scala:164)
        at org.scalafmt.dynamic.ScalafmtReflect.format(ScalafmtReflect.scala:98)
        at org.scalafmt.dynamic.ScalafmtDynamic.$anonfun$tryFormat$1(ScalafmtDynamic.scala:215)
        at scala.util.Try$.apply(Try.scala:213)
        at org.scalafmt.dynamic.ScalafmtDynamic.tryFormat(ScalafmtDynamic.scala:203)
        at org.scalafmt.dynamic.ScalafmtDynamic.$anonfun$formatDetailed$1(ScalafmtDynamic.scala:109)
        at scala.util.Either.flatMap(Either.scala:341)
        at org.scalafmt.dynamic.ScalafmtDynamic.formatDetailed(ScalafmtDynamic.scala:108)
        at org.scalafmt.dynamic.ScalafmtDynamic.format(ScalafmtDynamic.scala:70)
        at org.scalafmt.cli.ScalafmtDynamicRunner$.handleFile(ScalafmtDynamicRunner.scala:96)
        at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$4(ScalafmtDynamicRunner.scala:55)
        at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$4$adapted(ScalafmtDynamicRunner.scala:49)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at org.scalafmt.cli.ScalafmtDynamicRunner$.$anonfun$run$3(ScalafmtDynamicRunner.scala:49)
        at scala.util.control.Breaks.breakable(Breaks.scala:42)
        at org.scalafmt.cli.ScalafmtDynamicRunner$.run(ScalafmtDynamicRunner.scala:49)
        at org.scalafmt.cli.Cli$.runWithRunner(Cli.scala:135)
        at org.scalafmt.cli.Cli$.run(Cli.scala:84)
        at org.scalafmt.cli.Cli$.mainWithOptions(Cli.scala:69)
        at org.scalafmt.cli.Cli$.main(Cli.scala:58)
        at org.scalafmt.cli.Cli.main(Cli.scala)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at coursier.bootstrap.launcher.a.a(Unknown Source)
        at coursier.bootstrap.launcher.Launcher.main(Unknown Source)
```

### Cause

- That stacktrace comes from https://github.com/adampauls/scalafmt/blob/0b8b3da0b46b41767bedf3a059b29163c38f605f/scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtReflect.scala#L159-L166
  - If the version in `.scalafmt.conf` set to SNAPSHOT version, it always throws the VersionMismatchException because
    - ScalafmtReflect received parsed to `toString`ed `ScalafmtVersion` https://github.com/adampauls/scalafmt/blob/0b8b3da0b46b41767bedf3a059b29163c38f605f/scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtDynamic.scala#L156-L180
    - but [`ScalafmtVersion` doesn't have SNAPSHOT information](https://github.com/scalameta/scalafmt/blob/784d242f09f6c6aa83e6f2d129aa631cf24b762f/scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtVersion.scala#L5), as a result, `2.2.2-SNAPSHOT` down to `2.2.2`.

### Solution
To fix this, this PR

- enables `ScalafmtVersion` to have SNAPSHOT information
- if the version is SNAPSHOT, add `-SNAPSHOT` suffix on `toString` of `ScalafmtVersion`

### CODE REVIEW

File: scalafmt-dynamic/src/test/scala/tests/ScalafmtVersionSuite.scala [hunk 1]
```diff
@@ -68,16 +103,38 @@ class ScalafmtVersionSuite extends AnyFunSuite {
...
+    )
+    assert(
+      ScalafmtVersion(0, 1, 8, 0, false) < ScalafmtVersion(0, 2, 2, 0, false)
+    )
+    assert(
+      !(ScalafmtVersion(2, 0, 0, 0, false) < ScalafmtVersion(1, 5, 1, 0, false))
+    )
+    assert(
+      !(ScalafmtVersion(0, 1, 8, 0, false) > ScalafmtVersion(0, 2, 2, 0, false))
+    )
```
Comment (tanishiking) [inline scalafmt-dynamic/src/test/scala/tests/ScalafmtVersionSuite.scala]: According to semantic versioning spec, we might should define the order `2.2.2 > 2.2.2-SNAPSHOT > 2.2.1` or something.
However, considering the use-case of `ScalafmtVersion` in scalafmt-dynamic, we don't need to implement such versioning I guess :)

Comment (poslegm) [reply scalafmt-dynamic/src/test/scala/tests/ScalafmtVersionSuite.scala]: May be cover this case? It's not looking hard:

```scala
def <(other: ScalafmtVersion): Boolean =
    if (integerRepr == other.integerRepr && rc == other.rc)
      snapshot && !other.snapshot
    else if (integerRepr == other.integerRepr)
      rc != 0 && (other.rc == 0 || rc < other.rc)	   
    else integerRepr < other.integerRepr	  
```

**Comment by poslegm** [comment]
Thank you!

File: scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtVersion.scala [hunk 2]
```diff
@@ -2,7 +2,13 @@ package org.scalafmt.dynamic
 
 import scala.util.control.{NonFatal, NoStackTrace}
 
-case class ScalafmtVersion(major: Int, minor: Int, patch: Int, rc: Int) {
+case class ScalafmtVersion(
+    major: Int,
+    minor: Int,
+    patch: Int,
+    rc: Int,
+    snapshot: Boolean
```
Comment (kitbellew) [inline scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtVersion.scala]: optional: if you add `= false`, you won't need to change most of the tests :)

File: scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtVersion.scala [hunk 3]
```diff
@@ -2,7 +2,13 @@ package org.scalafmt.dynamic
...
+) {
   private val integerRepr: Int =
     major * 100 + minor * 10 + patch
```
Comment (kitbellew) [inline scalafmt-dynamic/src/main/scala/org/scalafmt/dynamic/ScalafmtVersion.scala]: optional: if you replace with something like:
```
private val integerRepr: Int =
  major << 20 | minor << 15 | patch << 10 | rc << 5 | (snapshot == null)
```
can you then simplify comparison as:
```
def <(other: ScalafmtVersion): Boolean
  integerRepr < other.integerRepr
```

File: scalafmt-dynamic/src/test/scala/tests/ScalafmtVersionSuite.scala (hunk 1 above)
Comment (kitbellew) [reply scalafmt-dynamic/src/test/scala/tests/ScalafmtVersionSuite.scala]: see suggestion above. if you do implement it, adding a test to cover your proposed case would be ideal.

**Comment by tanishiking** [comment]
Thank you for your feedback @poslegm @kitbellew !
I added the default value for `snapshot = false` as mentioned https://github.com/scalameta/scalafmt/pull/1656#discussion_r372740449

Sorry for the confusion, I didn't mean to propose implementing order between snapshot version and not. We can implement it, but I don't see the advantage of that feature, so let me merge this. Thank you!

**Pull request merged**